### ISSUE TITLE
No easy way to intercept failed connection from pb.PBClientFactory

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10702]
|<img alt="lstep's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| lstep reported|
|-|-|
|Trac ID|trac#1049|
|Type|enhancement|
|Created|2005-05-24 16:03:39Z|
|Branch|https://github.com/twisted/twisted/tree/pb-login-connect-failure-1049|

<details><summary>Searchable metadata</summary>

```
trac-id__1049 1049
type__enhancement enhancement
reporter__lstep lstep
priority__normal normal
milestone__ 
branch__branches_pb_login_connect_failure_1049 branches/pb-login-connect-failure-1049
branch_author__ 
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__ 
time__1116950619000000 1116950619000000
changetime__1204235180000000 1204235180000000
version__ 
owner__ 
cc__exarkun cc__warner cc__lstep
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="lstep's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|lstep commented|
|-|-|

```
#!html
<pre>
In pb.Broker, there's a notifyOnFail() for failed connections, but it's only 
for the server side. On the client side (pb.PBClientFactory), there's 
notifyOnDisconnect(), that's all. 
That would be nice to be able to add a method that gets called when the 
reactor.connect* fails to connect. There's even a method for this, but to 
access it you need to subclass pb.PBClientFactory(). So something like: 
 
self.factory = pb.PBClientFactory() 
self.factory.notifyOnFail(self.couldNotConnect) 
reactor.connectTCP(ipSource, 8890, self.factory)
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
If you use PBClientFactory.login(), you should get a Deferred which errors when
the connection fails.  Is this insufficient?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="lstep's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|lstep commented|
|-|-|

```
#!html
<pre>
Hmm, no, that's what I'm doing, I have a  
def1.addCallbacks(callback=self.connected, errback=self.noLogin) and that  
self.noLogin ONLY catches login errors, but not connection errors. For  
example, if the distant machine does not listen on the specified port, the  
exception is not catched by the self.noLogin method.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Weak.  Would your use-case be satisfied if it actually *did* work the way I
described?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="lstep's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|lstep commented|
|-|-|

```
#!html
<pre>
On Friday 27 May 2005 17:22, Jp Calderone [Twisted issue tracker] wrote:
> Jp Calderone &lt;exarkun@...> added the comment:
>
> Weak.  Would your use-case be satisfied if it actually *did* work the way I
> described?

It's a solution, but the problem with it is that I will not able to know that 
the problem is from a login error or from a connection error (I mean 
network).
I thought it would be nicer to have some symmetry with the available client 
methods where there is are 'notifyOn*' that allows one to hook internal 
methods.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Even though you'll get a different kind of failure in each case?

You should get either a DNSLookupError or a ConnectError if the connection
simply cannot be established, but an UnauthorizedLogin or something else if
there is a problem authenticating.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [22716]) Branching to 'pb-login-connect-failure-1049'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun removed owner|
|-|-|

As I suspected, this seemed to work already.  I didn't see a unit test for it, though.  So, I added one in the branch.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to @exarkun|
|-|-|

The first argument of clientFactory.clientConnectionFailed should be None instead of object() (I think this is how we use to indicate an unused argument). The rest looks perfect, so please merge once done.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [22719]) Merge pb-login-connect-failure-1049

Author: exarkun
Reviewer: therve
Fixes [#10702](https://github.com/twisted/twisted/issues/10702)

Add a test for connection failure resulting the Deferred returned by PBClientFactory.login
being errbacked.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
